polar-core/src/partial/simplify.rs

```diff
@@ -1,217 +1,128 @@
-                    }
-                    (_, _) => panic!("invalid"),
+                    _ => fold_term(t, self),
                 }
-            } else {
-                args.push(term!(operation.clone()))
             }
+            _ => fold_term(t, self),
         }
-
-        operations.push(Operation {
-            operator: Operator::And,
-            args,
-        });
-    }
-}
-
-fn not_this_arg(operation: &Operation) -> Option<Term> {
-    let left = operation.args.get(0).unwrap();
-    let right = operation.args.get(1).unwrap();
-
-    match (is_this_arg(left.value()), is_this_arg(right.value())) {
-        (false, true) => Some(left.clone()),
-        (true, false) => Some(right.clone()),
-        _ => None,
     }
-}
-
-fn is_this_arg(value: &Value) -> bool {
-    matches!(value, Value::Variable(sym) if sym.0 == "_this")
-}
-
-// partial(_x_5) { partial(_value_1_6) { _this > 0, _this > 1 } = _this.a }
 
-// Take partial(_this = ?) and output ?.
-fn simplify_unify_partials(term: Term, _: &Bindings) -> Term {
-    if let Value::Partial(p) = term.value() {
-        let operator = p.operations().first().unwrap().operator;
-        let is_unify = matches!(operator, Operator::Unify);
-
-        if p.operations().len() == 1 && is_unify {
-            let op = p.operations().first().unwrap();
-
-            match not_this_arg(op) {
-                Some(term) => term,
-                None => term.clone(),
+    fn fold_operation(&mut self, o: Operation) -> Operation {
+        /// Given `_this` and `x`, return `x`.
+        /// Given `_this.x` and `_this.y`, return `_this.x.y`.
+        fn sub_this(term: &Term, replacement: &Term) -> Term {
+            match (term.value(), replacement.value()) {
+                (Value::Variable(v), _) if v.is_this_var() => replacement.clone(),
+                (
+                    Value::Expression(Operation {
+                        operator: Operator::Dot,
+                        args,
+                    }),
+                    Value::Expression(Operation {
+                        operator: Operator::Dot,
+                        ..
+                    }),
+                ) => term.clone_with_value(Value::Expression(Operation {
+                    operator: Operator::Dot,
+                    args: vec![replacement.clone(), args.get(1).unwrap().clone()],
+                })),
+                _ => term.clone(),
             }
-        } else {
-            term.clone()
         }
-    } else {
-        term.clone()
-    }
-}
 
-fn get_roots(bindings: &Bindings) -> HashSet<Symbol> {
-    let mut roots = HashSet::new();
-    for (symbol, val) in bindings.iter() {
-        if !symbol.is_temporary_var() {
-            if let Value::Partial(_) = val.value() {
-                roots.insert(symbol.clone());
+        // Optionally sub `replacement` into each of the arguments of the operations.
+        let mut map_ops = |ops: &[Operation], replacement: &Term| -> TermList {
```

Comment:
Could we name the arguments here `partial_ops` and `expression`? I think that would make it easier to go from the match statement to reading this function.

Same for `sub_this`.

Reply:
Some comments would be helpful as well.

Reply:
Renamed in c3638f2 and added a couple comments.